alf/networks/encoding_networks.py

```diff
@@ -0,0 +1,314 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gin
+import numpy as np
+
+import torch
+import torch.nn as nn
+
+import alf.layers as layers
+
+from alf.tensor_specs import TensorSpec
+
+
+def _tuplify2d(x):
+    if isinstance(x, tuple):
+        assert len(x) == 2
+        return x
+    return (x, x)
+
+
+@gin.configurable
+class ImageEncodingNetwork(nn.Module):
+    """
+    A general template class for creating convolutional encoding networks.
+    """
+
+    def __init__(self,
+                 input_channels,
+                 conv_layer_params,
+                 activation=torch.relu,
+                 flatten_output=False):
+        """
+        Initialize the layers for encoding an image into a latent vector.
+
+        Args:
+            input_channels (int): number of channels in the input image
+            conv_layer_params (list[tuple]): a non-empty list of elements
+                (num_filters, kernel_size, strides, padding), where padding is
+                optional
+            activation (torch.nn.functional): activation for all the layers
+            flatten_output (bool): If False, the output will be an image
+                structure of shape `BxCxHxW`; otherwise the output will be
+                flattened into a feature of shape `BxN`
+        """
+        super(ImageEncodingNetwork, self).__init__()
+
+        assert isinstance(conv_layer_params, list)
+        assert len(conv_layer_params) > 0
+
+        self._flatten_output = flatten_output
+        self._conv_layer_params = conv_layer_params
+        self._conv_layers = nn.ModuleList()
+        for paras in conv_layer_params:
+            filters, kernel_size, strides = paras[:3]
+            padding = paras[3] if len(paras) > 3 else 0
+            self._conv_layers.append(
+                layers.Conv2D(
+                    input_channels,
+                    filters,
+                    kernel_size,
+                    activation=activation,
+                    strides=strides,
+                    padding=padding))
+            input_channels = filters
+
+    def output_shape(self, input_size):
```

Comment:
I'm wondering, can we pass a 0 tensor of input_size to forward() and get the output_size from the output tensor?

That'll probably reduce code redundancy and won't add much overhead?

If we do this, then we can have a very simple output_shape function in the top level class, and all our custom nn.Module's can just inherit without explicitly implementing it?

Reply:
Calculating output_shape is usually required for conv networks because for other layers you will know its value during construction. As for why I didn't use forward to get the output_shape is mainly for testing purpose in encoding_networks_test.py. The code amount won't be much less by doing forward.

Reply:
I see.  There are already two Modules that need output shape in this PR.  I worry less about code length, and more about code maintenance in the future.  If using forward can solve the problem once and for all, then why not?

Reply:
Because for encoding_networks_test.py, I compare the forwarding-computed shape with the maunally calculated one. It's basically a sanity check.

Reply:
Then that means we can remove that test if we use forward to compute output shape in the output_shape() function?

Reply:
Then how do you check if the networks are implemented correctly?

Reply:
Maybe move the manual computation logic into the test if you want.  Otherwise, just get a few example gold standard output shapes into the test?

Reply:
The only downside I can think of is whenever output_shape() is called, especially when input size is large, gpu memory could rocket.  But it's hopefully just a one time cost.

Reply:
Also I personally don't feel comfortable that we are doing forward under the hood without telling the users. The log (error) message, if any, will be quite confusing, just like what we have now for prepare_specs.

Reply:
Ok, fair enough.  We can always reconsider if needed.